# 5-1-2. return 없는 Closure

> 클로저<br>: 어떤 함수 A에서 선언한 변수 a를 참조하는 내부함수 B를 외부로 전달할 경우<br>**A의 실행 컨텍스트가 종료된 이후에도 변수 a가 사라지 않는 현상**
> 

일반 `중첩함수`를 통해 `외부 함수에서 내부 함수를 return 하는 경우`를 클로저 예시로 많이 볼 수 있습니다.

하지만 **return 없이도 클로저가 발생하는 경우**도 있습니다.

```jsx
let outer = function() {
	let a = 0;
	let intervalId = null;

	let inner = function () {
		if (a >= 10) {
			clearInterval(intervalId);
		}
		a++;
		console.log(a);
	}

	intervalId = setInterval(inner, 1000);
}

outer();
```

- **내부 함수 inner**에서 **outer 함수**에 선언한 변수 **a**를 참조하고 있습니다.
- outer함수의 실행 컨텍스트가 종료되었는데도 내부 함수 inner는 여전히 동작하고 있습니다.
(`외부 함수(outer)`보다 `중첩 함수(inner)`가 더 오래 유지되는 상황)
- 그래서 여전히 **변수 a**는 **inner 함수에서의 참조**로 살아있는 경우이기에, `클로저`가 발생하는 상황이라고 할 수 있습니다.

<br>

```jsx
let outer = function() {
	let count = 0;
	let button = document.createElement('button');
	button.innerText = 'click';
	
	button.addEventListener('click', function() {
		console.log(count, 'times clicked');
		count++;
	});
	document.body.appendChild(button);
}

outer();
```

- **outer 함수 내부**에서 **addEventListener 메서드**를 통해 콜백 함수를 호출하고 있습니다.
- **outer 함수의 실행 컨텍스트**는 종료되었지만, 여전히 **addEventListener 메서드**는 동작하고 있습니다.
- 그래서 **addEventListener 메서드의 콜백함수**에서 참조하는 **outer 함수에 선언된 변수 count**는 죽지 않고 살아있습니다.
- 이때도 `클로저`가 발생하는 상황이라고 할 수 있습니다.
